pkg/util/grpc.go

```diff
@@ -45,6 +83,272 @@ func StatusCode(err error) codes.Code {
+}
+
+func shouldGrpcRevalidate(md metadata.MD, cacheType string) bool {
+	for _, v := range md.Get(GrpcCacheControlHeader(cacheType)) {
+		if v == GrpcNoCache {
+			return false
+		}
+		if v == GrpcMustRevalidateCache {
+			return true
+		}
+	}
+	return false
+}
+
+func sign(v, signature string) string {
+	return fmt.Sprintf("%s_%s", v, signature)
+}
+
+func signMetadata(md metadata.MD, fullMethod string) {
+	vals := md.Get(GrpcCacheControlHeader(CacheTypeClient))
+	for i, v := range vals {
+		if strings.Contains(v, "max-age") {
+			vals[i] = sign(v, fullMethod)
+		}
+	}
+	md.Set(GrpcCacheControlHeader(CacheTypeClient), vals...)
+}
+
+func getCacheControlDetails(md metadata.MD, cacheType, signature string) (time.Duration, string) {
+	for _, v := range md.Get(GrpcCacheControlHeader(cacheType)) {
+		if strings.Contains(v, "max-age") {
+			parts := strings.Split(v, "=")
+			if len(parts) != 2 {
+				continue
+			}
+			info := strings.Split(parts[1], "_")
+			if len(info) != 2 {
+				continue
+			}
+			ttl, err := strconv.Atoi(info[0])
+			if err == nil && info[1] == signature {
+				return time.Duration(ttl) * time.Second, info[1]
+			}
+		}
+	}
+	return 0, ""
+}
+
+func checkSignature(v, signature string) bool {
+	headerComponents := strings.Split(v, "_")
+	if len(headerComponents) == 2 {
+		return headerComponents[1] == signature
+	}
+	return false
+}
+
+func shouldGrpcCache(md metadata.MD, cacheType, signature string) bool {
+	if md == nil {
+		return false
+	}
+	shouldCache := false
+	for _, v := range md.Get(GrpcCacheControlHeader(cacheType)) {
+		if v == GrpcNoCache {
+			return false
+		}
+		if strings.Contains(v, "max-age") {
+			if checkSignature(v, signature) {
+				shouldCache = true
+			}
+		}
+	}
+	return shouldCache
+}
+
+func isCacheMiss(md metadata.MD, cacheType string) bool {
+	for _, v := range md.Get(GrpcCacheStatusHeader(cacheType)) {
+		if v == GrpcCacheMiss {
+			return true
+		}
+	}
+
+	return false
+}
+
+func key(method string, req proto.Message) string {
+	val := reflect.Indirect(reflect.ValueOf(req))
+	typ := val.Type() // get the type of the underlying value
+	if reflect.PtrTo(typ).Implements(reflect.TypeOf((*caching.CacheKeyer)(nil)).Elem()) {
+		return req.(caching.CacheKeyer).CacheKey()
+	}
+	return hash(method, req)
+}
+
+type GrpcCachingInterceptor interface {
+	UnaryServerInterceptor() grpc.UnaryServerInterceptor
+	UnaryClientInterceptor() grpc.UnaryClientInterceptor
+	storage.GrpcTtlCache
+}
+
+type GrpcClientEntityCacher struct {
+	storage.GrpcTtlCache
+	cacheType string
+}
+
+var _ GrpcCachingInterceptor = (*GrpcClientEntityCacher)(nil)
+
+func NewClientGrpcEntityCacher(cache storage.GrpcTtlCache) *GrpcClientEntityCacher {
+	return &GrpcClientEntityCacher{
+		GrpcTtlCache: cache,
+		cacheType:    CacheTypeClient,
+	}
+}
+
+// both the server and client agree on a signature that identifies the RPC that should be cached
+func (g *GrpcClientEntityCacher) sign(signature string) metadata.MD {
```

Comment:
I would use different terminology here, "signature" refers to a very specific kind of thing.